docs/workflow-semantics.md

```diff
@@ -77,19 +61,22 @@ each|iterates over all elements of a hash or an array|`collection` - a hash or a
 An activity with an iteration must be declared to produce an `output` consisting of a `key` and a `value`. This is the product of each iteration. The final output from the activity is one variable named after the activity containing a Hash consisting of the key/value pair of each iteration.
 
 ### Action
-An action defines imperative code that performs some tasks related to a common subject or subjects. The `<data or code>` block is divided into three blocks of code named `delete`, `read`, and `upsert`. A missing block means that the action does not take part in the corresponding operation.
+An action defines a set of functions that performs tasks related to a common subject or subjects. An action is often a specialization of some type of handling necessary for other types of activities. At present, the only relevant examples of an actions are the CRUD handlers for resource activities.
 
-The `delete` block will not have access to dynamic input (i.e. input produced by output from other activities) and it is not supposed to provide any output. It will have access to workflow variables that stem from lookups or was provided to the workflow engine as initial parameters.
+### Stateless
+Stateless defines imperative code that performs some tasks based on its input and provides some output. A Stateless activity will typically consult the `context` to determine what operation to perform during an apply.
```

Comment:
I really don't understand the difference between `action` and `stateless`; as an end-user, when do I choose one over another?

Reply:
I agree with @jdwelch. I know we intentionally architected the workflow engine so that a "provider" is just another type of plugin. However, `action` as currently defined to me is a special interface for someone creating types/providers and not end-users. 

My preference would be for something like:
`resource_provider` describes the CRUD operations required on a resource
`action` describes imperative code that performs some tasks. 
`resource` describes declarative resource (as we know today)

Reply:
+1. Will file a naming discussion ticket shortly.